oshmem/mca/scoll/ucc/scoll_ucc_module.c

```diff
@@ -51,17 +49,21 @@ int mca_scoll_ucc_progress(void)
 
 static void mca_scoll_ucc_module_destruct(mca_scoll_ucc_module_t *ucc_module)
 {
-    ucc_team_destroy(ucc_module->ucc_team);
+    if (ucc_module->ucc_team) {
+        ucc_team_destroy(ucc_module->ucc_team);
+        --mca_scoll_ucc_component.nr_modules;
+    }
 
-    if (ucc_module->group->ompi_comm == (ompi_communicator_t *) &oshmem_comm_world) {
-        if (mca_scoll_ucc_component.libucc_initialized) {
+    if (1 == mca_scoll_ucc_component.nr_modules) {
```

Comment:
HOw will this work if the sequence is:
1st group is create (nr_modules = 1, initialized = true)
2nd group is created (nr_modules = 2)
2nd group destroyed (nr_modules becomes 1 and ucc is finalized)

now 3rd group is created. -> will it fail?

Reply:
As of the supported OpenSHMEM specification, groups are created whenever a unique active set is used for a collective and groups are cached until shmem_finalize(), where groups will be destroyed in the order they were created. 
 
The stopping condition is set to 1 instead of 0 because there is an initial call to initialize the component and enable a UCC module during oshmem_proc_group_create(), which is then immediately destroyed. I put some checks in place lower in the code to check for this and skip it as it is multiple unnecessary allgathers. 

As there is no mechanism in place for a user to destroy an osh_group, the sequence would look like with 3 groups:

1st group skipped and is not used (nr_modules = 1, initialized = false)
1st group destroyed (nr_modules = 1, initialized = false)
2nd group created (nr_modules = 2, initialized = true) // this is oshmem_group_all
3rd group created (nr_modules = 3)
2nd group destroyed (nr_modules = 2)
3rd group destroyed (nr_modules = 1 and ucc is finalized)

Reply:
@vspetrov Does this make sense?

Reply:
i think so. when you say "supportec OpenSHMEM specification" - could you point me to the spec description of groups being cached until shmem_finalize?

Reply:
I phrased that wrong I believe. What I meant was that the supported version of OpenSHMEM in oshmem is 1.4, which is why the collectives are using active sets rather than teams. 

The code I was referring to that invokes this code is in oshmem_proc_group_finalize_scoll() in oshmem/proc/proc.c.

Reply:
got it, thanks.